File: leetcode/0906-super-palindromes/Solution.java
class Solution {
    public int superpalindromesInRange(String left, String right) {
        long lo = Long.parseLong(left);
        long hi = Long.parseLong(right);
        int count = 0;
        int limit = 100000;

        for (int seed = 1; seed < limit; seed++) {
            String s = Integer.toString(seed);

            String oddStr = s + new StringBuilder(s.substring(0, s.length() - 1)).reverse().toString();
            String evenStr = s + new StringBuilder(s).reverse().toString();

            long oddRoot = Long.parseLong(oddStr);
            long evenRoot = Long.parseLong(evenStr);

            for (long root : new long[]{oddRoot, evenRoot}) {
                long square = root * root;
                if (square > hi) continue;
                if (square >= lo && isPalindrome(square)) {
                    count++;
                }
            }
        }

        return count;
    }

    private boolean isPalindrome(long n) {
        String s = Long.toString(n);
        int i = 0, j = s.length() - 1;
        while (i < j) {
            if (s.charAt(i) != s.charAt(j)) return false;
            i++;
            j--;
        }
        return true;
    }
}

File: leetcode/0906-super-palindromes/solution.cpp
#include <string>
#include <algorithm>
using namespace std;

class Solution {
public:
    int superpalindromesInRange(string left, string right) {
        long long lo = stoll(left);
        long long hi = stoll(right);
        int count = 0;
        int limit = 100000;

        for (int seed = 1; seed < limit; seed++) {
            string s = to_string(seed);

            string oddRev = s.substr(0, s.size() - 1);
            reverse(oddRev.begin(), oddRev.end());
            string oddStr = s + oddRev;

            string evenRev = s;
            reverse(evenRev.begin(), evenRev.end());
            string evenStr = s + evenRev;

            long long oddRoot = stoll(oddStr);
            long long evenRoot = stoll(evenStr);

            for (long long root : {oddRoot, evenRoot}) {
                long long square = root * root;
                if (square > hi) continue;
                if (square >= lo && isPalindrome(square)) {
                    count++;
                }
            }
        }

        return count;
    }

private:
    bool isPalindrome(long long n) {
        string s = to_string(n);
        int i = 0, j = (int)s.size() - 1;
        while (i < j) {
            if (s[i] != s[j]) return false;
            i++;
            j--;
        }
        return true;
    }
};
